lib/spack/spack/cmd/develop.py

```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-        if args.build_directory is not None:
+        if build_dir is not None:
             spack.config.add(
-                "packages:{}:package_attributes:build_directory:{}".format(
-                    spec.name, args.build_directory
-                ),
+                f"packages:{spec.name}:package_attributes:build_directory:{build_dir}",
                 env.scope_name,
             )
-        _update_config(spec, path)
+        # add develop spec and update path
+        if specified_path:
+            # always overwrite path if it was supplied
+            _update_config(spec, specified_path)
+        else:
+            _update_config(spec, spec.name)
+
+
+def _clone(spec: spack.spec.Spec, abspath: str, force: bool = False):
+    if os.path.exists(abspath):
+        if force:
+            shutil.rmtree(abspath)
+        else:
+            msg = f"Skipping developer download of {spec.name}"
+            msg += f" because its path {abspath} already exists."
+            tty.msg(msg)
+            return
+
+    # cloning can take a while and it's nice to get a message for the longer clones
+    tty.msg(f"Cloning source code for {spec}")
+    _retrieve_develop_source(spec, abspath)
+
+
+def _abs_code_path(
+    env: spack.environment.Environment, spec: spack.spec.Spec, path: Optional[str] = None
+):
+    src_path = path if path else spec.name
+    return spack.util.path.canonicalize_path(src_path, default_wd=env.path)
+
+
+def _dev_spec_generator(args, env):
+    """
+    Generator function to loop over all the develop specs based on how the command is called
+    If no specs are supplied then loop over the develop specs listed in the environment.
+    """
+    if not args.spec:
+        if args.clone is False:
+            raise SpackError("No spec provided to spack develop command")
+        else:
+            for name, entry in env.dev_specs.items():
+                path = entry.get("path", name)
+                abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
+                # Both old syntax `spack develop pkg@x` and new syntax `spack develop pkg@=x`
+                # are currently supported.
+                spec = spack.spec.parse_with_version_concrete(entry["spec"])
+                yield spec, abspath
+    else:
+        specs = spack.cmd.parse_specs(args.spec)
+        if (args.path or args.build_directory) and len(specs) > 1:
+            raise SpackError(
+                "spack develop requires at most one named spec when using the --path or"
+                " --build-directory arguments"
+            )
+        else:
+            for spec in specs:
+                if args.recursive:
+                    concrete_specs = env.all_matching_specs(spec)
+                    if not concrete_specs:
+                        tty.msg(
+                            "No matching specs found in the environment. "
+                            "Recursive develop requires a concretized environment"
```

Comment:
This needs to say which spec didn't match, and should make clear that this spec is being skipped but others will still be developed.